### ISSUE TITLE
Performance tuning

### ISSUE THREAD

**Issue by skleinfeldt** [jazkarta/jazkarta.com#10]
The site is reeeeeaaly slow. Can we get better performance without upgrading instance type?

**Comment by skleinfeldt** [comment]
Hold this until after #13, which might solve the problem.

**Comment by skleinfeldt** [comment]
@alecpm can you take a look to see if there are any quick improvements that could be made? When logged in every page takes several seconds to load which does slow down the editing work.

This is not worth spending much time on, just a few minutes to see if there are any obvious problems.

**Comment by alecpm** [comment]
Turning on some basic plone.app.caching config to allow varnish (and the browser cache) to do its job seems to have helped.  We might want to try a t2.medium instance, which is a little cheaper but has more RAM and is built for burst-y applications (so it can make a lot of sense for a staging server).

**Comment by skleinfeldt** [comment]
Anonymous viewing does seem better. Logged in, this morning it's noticeably worse for me than late last night (as measured by the subjective counting seconds by "1-a-and-y, 2-a-and-y" method), but I guess that must be my internet connection. So I should do my content editing in the wee hours...

If you'd like to try out the t2.medium to see how it behaves, that would be fine with me. Up to you.

We will be continuing to do frequent deployments and restarts for a while, even after launch I expect, and I think I have been noticing the cold cache problem. Would it be worth opening a ticket to install a cache-filling add-on that runs on startup? (I think I've read there are things like that.)

**Comment by alecpm** [comment]
I don't think such an add-on exists currently, though I know I've head talk of custom solutions for filling the object cache on startup.  I'm not sure it would be a huge win for this site though.  I'll try out one of the other instance types tomorrow.

**Comment by skleinfeldt** [comment]
OK, let me know when you change the instance type I'll be curious if it seems to help.

**Comment by skleinfeldt** [comment]
Closing this, assume we don't want to take any more time on this for now, and I am getting by through the miracle of multiple editing tabs.

**Comment by alecpm** [comment]
I'm still planning to do a t2.medium upgrade, but I don't think we need a ticket for that.  The slowness is very curious though.

**Comment by alecpm** [comment]
I'm going to reopen this because there are some perplexing performance quirks.  CPU and RAM use seems pretty low on average.  We're running on a t2.medium now, and the CPU credit balance is stable over the past week (which basically means we have more CPU than we need) for the current workload.

Nonetheless, things just seem to stall somewhat randomly, even SSH connections seem to sometimes take forever to negotiate.  Looking at iostat, it isn't showing high I/O usage or notable iowait times.  Newrelic is showing fairly low average server response times, which may mean the delays are happening on the connection side before any python code runs.  New Relic is showing that the cropping editor is taking up a lot of page load time, but that's somewhat expected since it's an edit view that's performing image transformations and @skleinfeldt is apparently using it a lot.  The New Relic browser monitor is showing some pretty significant network delays.

I may try using a tmpfs filesystem for /tmp (and maybe the plone instance tmp), in case I/O to /tmp is part of the problem.  Traditionally we've run our /tmp off of instance store storage which has tended to have more consistent I/O performance than EBS, but the new instance types are mostly EBS only.  However, given that the issues appear to not be on the application server side, but on the network side, I suspect that won't help much.  I'll continue looking into this, and if there's a fix that can be incorporated into our deployment recipes I'll do so.

**Comment by cguardia** [comment]
Would you say that is a blocker, or do we go on with the launch for today?

**Comment by alecpm** [comment]
No, that's why I put it in the Later milestone.  I'll continue to work on it a bit, but we should be good to go.  Anon page views are mostly performing decently.

**Comment by alecpm** [comment]
It seems like something has improved performance quite a bit.  I added a tmpfs at /tmp, but it appears writes there are pretty infrequent (mainly just some newrelic logging).  I implemented some sysctl config changes based on these recommendations from Netflix (http://www.brendangregg.com/blog/2015-03-03/performance-tuning-linux-instances-on-ec2.html), though I have yet to incorporate those into the deploy recipes.

I also made some customizations to our cache ruleset configuration to ensure some resources that weren't being cahed are now, but those were made after the performance improvements I'm seeing.  The improvement could just be random variance in network performance or something though.
